### PR TITLE
Interpreter integration

### PR DESCRIPTION
**Pull request by ftc**
Work in progress

TODO:
We need to wrap the following objects and implement a statement interpreter:
- [x] JimpleBody
- [x] soot.Local
- [x] soot.JimpleLocal
- [x] JAssignStmt
- [x] JReturnStmt

I have implemented a stub interpretBody function which shows how these can be used, we will probably need to replace it once all the things are wrapped but it works for a function of "return 2" and there is a second unit test with a local in there.

### CODE REVIEW

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (bechang) [inline src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Why are sucessors taken in the JAssignStmt here?

Comment (bechang) [reply src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Versus some Seq statement?

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (bechang) [inline src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: You will have an unapply for getLeftOp and getRightOp/

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (bechang) [reply src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Control-flow processing should be "outside".

**Comment by ftc** [comment]
Status: @jpw234 is finishing the last two wrapped objects then I believe this task should be done.  I added a bunch of unit tests for various arithmetic operations.

**Comment by ftc** [comment]
I believe this is ready for code review.

**Comment by ftc** [comment]
@bennostein if you could review this I would appreciate it.  There is some nastiness in the control flow right now, specifically the lines:

```
63    val successor: Set[Stmt] = b.getSuccessors(loc)
```
This is just a way of finding the next nodes to be processed, however as I discovered implementing GotoStmt and IfStmt they have their own "target" fields.



```
76        if(successor.size == 1)
75          internal_interpretBody(newEnv, successor.iterator.next(), b)
76        else
77          ??? //malformed
```
Some nodes shouldn't have more than one successor.  I think we probably want to refactor the successor mechanism in general which will get rid of this.

both of these are in the interpreter.scala here:

https://github.com/cuplv/cuanto/blob/interpreter-integration/src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala

I propose we accept this as is and then iterate on control flow stuff with the next issue #68 which we can talk about this Friday.

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (bennostein) [inline src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Should this be `private` or `protected` if it's internal?

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (bennostein) [inline src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Will look at this successor thing more closely when I'm out of class.

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (bechang) [inline src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Get rid of this junk stuff before merge

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (smover) [inline src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: This should be an assertion

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (bechang) [inline src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Shouldn't use .get

**Review by bennostein — APPROVED** [review]
This looks good to me.  I agree that we should implement something more robust/general for finding successors, but think that this is a big enough chunk already that that should be a separate PR.

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (ftc) [reply src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: Changed to throw 
```
.getOrElse(throw new RuntimeException("interpreter exception")
```
This allows some information to be dumped into the assertion failure.  We should discuss how best to report unexpected failures but this is a pattern that I have used in the past.

File: src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala
Comment (ftc) [reply src/main/scala/edu/colorado/plv/cuanto/scoot/concrete_interpreter/Interpreter.scala]: I made everything private that shouldn't be used by outside users of the interpreter except the method "evaluate_expr(v: Value, env: Env = emptyEnv): Option[Int]" since it needs to be tested.  We should look into a lower level of accessibility that lets it be unit tested but not externally utilized.

**Comment by ftc** [comment]
@bennostein I implemented the suggestions from you, @bechang, and @smover. I think we should look at another round of code review.  I put some specific comments on the comments that existed before.

One place I would like feedback on is the following function as I am not convinced this is the cleanest way to do it but it seemed to deduplicate a bunch of code.

```
  private def wrapExprEvaluationException(exprResult: Option[Int],
                                          stmt: Stmt,
                                          successCondition: Int => NormalControlFlow): StmtResult = exprResult match{
    case Some(v) => successCondition(v)
    case None => ExecutionExceptionDivideByZero(stmt)
  }
```

**Comment by bennostein** [comment]
Looks like my "approve" still hold even after these new changes, but just to be clear: I approve all of the new changes, following our discussions today!

**Pull request merged**
